File: Python/BOJ/BOJ20303.py
def find(a):
    if a != parents[a]:
        parents[a] = find(parents[a])
    return parents[a]

def union(a, b):
    pa = find(a)
    pb = find(b)
    if pa == pb:
        return
    if size[pa] > size[pb]:
        parents[pb] = pa
        size[pa] += size[pb]
        c[pa] += c[pb]
    else:
        parents[pa] = pb
        size[pb] += size[pa]
        c[pb] += c[pa]

n, m, k = map(int, input().split())
c = [0] + list(map(int, input().split()))
parents = list(range(n + 1))
size = [1] * (n + 1)
for _ in range(m):
    a, b = map(int, input().split())
    union(a, b)

dp = [0] * k
for i in range(1, n + 1):
    if i == find(i):
        w = size[i]
        v = c[i]
        for j in range(k - 1, w - 1, - 1):
            dp[j] = max(dp[j], dp[j - w] + v)

print(max(dp))

File: cpp/boj/boj20303.cpp
#include <iostream>
#include <algorithm>

int find(int a, std::vector<int> &parents) {
    if (a != parents[a]) parents[a] = find(parents[a], parents);
    return parents[a];
}

void unionSets(int a, int b, std::vector<int> &parents, std::vector<int> &size, std::vector<int> &c) {
    int pa = find(a, parents);
    int pb = find(b, parents);
    if (pa == pb) return;
    if (size[pa] > size[pb]) {
        parents[pb] = pa;
        size[pa] += size[pb];
        c[pa] += c[pb];
    } else {
        parents[pa] = pb;
        size[pb] += size[pa];
        c[pb] += c[pa];
    }
}

int main() {
    int n, m, k, answer = 0;
    std::cin >> n >> m >> k;
    std::vector<int> c(n + 1, 0);
    std::vector<int> parents(n + 1, 0);
    std::vector<int> size(n + 1, 1);
    for (int i = 1; i <= n; i++) {
        std::cin >> c[i];
        parents[i] = i;
    }
    for (int i = 0; i < m; i++) {
        int a, b;
        std::cin >> a >> b;
        unionSets(a, b, parents, size, c);
    }
    std::vector<int> dp(k, 0);
    for (int i = 1; i <= n; i++) {
        if (i != find(i, parents)) continue;
        int w = size[i];
        int v = c[i];
        for (int j = k - 1; j >= w; j--) dp[j] = std::max(dp[j], dp[j - w] + v);
    }
    for (int i = 0; i < k; i++) answer = std::max(answer, dp[i]);
    std::cout << answer << std::endl;
    return 0;
}
